300 offline handling

kasperosterbye opened the pull request:
This PR prepares Microdown for BeautifulComments to do its Offline handling

kasperosterbye commented:
Not ready yet - working in tandem with a BeautifulComment PR to make sure they play together.

kasperosterbye commented:
Ready to roll. When you have time @Ducasse. It is not a showstopper for me to get this one merged. It works with Beautiful comments PR #14, and gives Offline settings to Beautiful Comments.

Ducasse commented on src/Microdown/MicHTTPResource.class.st:
```diff
@@ -34,16 +35,24 @@ MicHTTPResource class >> imageCache [
-	on: Error
-	do: [ ^nil ]
+	on: Exception
+	do: [ ^ nil ]
```
I was thinking that we could just return an image so that the method always returns polymorphic objects.

kasperosterbye replied:
The nil is used to let the renderer rather than the ressource decide what to do when no image is found. I believe this is the best way.

kasperosterbye replied:
I am not convinced the polymorphism should sit here. I will leave it as is.

Ducasse commented on src/Microdown/MicHTTPResource.class.st:
```diff
@@ -7,24 +7,25 @@ Class {
+		ifNil: [
+			^ self imageCache at: aUrl put: ((self rawImageFrom: aUrl) ifNil: [ ^nil ])]
+		ifNotNil: [ 
+			^ ImageNotCachedBlock cull: aUrl]
```
Why do you use cull: and not value: because cull: is costly and reflective so I really tend to use it only when absolutely needed.

kasperosterbye replied:
I believe it was a leftover because I allowed to use a block that could just return a image without using a url - like the box with a "?"

kasperosterbye replied:
I changed it to use value. I had not looked into cull.

Ducasse commented on src/Microdown/MicHTTPResource.class.st:
```diff
@@ -7,24 +7,25 @@ Class {
+	(self imageCache includesKey: aUrl)
+		ifTrue: [ ^ self imageCache at: aUrl ].
+		
+	ImageNotCachedBlock 
```
I found the logic with nil instead of boolean for ImageNotCachedBlock difficult to follow. 
I could not tell you if ifNil is caching or not. 
BTW you do not need the return in the ifNil: and ifNotNil: block

kasperosterbye replied:
I tend to agree that is a bit strange - I tried to tell what it was by adding "Block" at the end. It is a way in which HTTP resource avoids supporting Beautiful comments directly, but to give BC the functionality it need to do its offline stuff. The actual block is set in BC as part of its Offline handling, and is nil by default.

kasperosterbye replied:
I will look at the ifNil thingie

Ducasse commented:
Thanks Kasper!!!!

kasperosterbye commented:
OMG - that was a lot of interesting errors in the dev branch. Happy hacking @Ducasse!

kasperosterbye commented:
@Ducasse @tesonep - I fixed the dependency on PillarCore (and reverted MicRichTextComposer to use the stylers again). 
I did that in PR #319. I merged it myself wanting to see if it would indeed fix the CI on the other failing PR's.
Now the PR's are green.
